File: dsa/kmp.py
# O(n + m)
def kmp(self, elems: str, looking_for: str) -> int:
    # kmp
    def matcher(string, pattern):
        N = len(string)
        M = len(pattern)
        lsp = get_lsp(pattern)
        to_match = 0

        for end in range(N):
            while to_match > 0 and string[end] != pattern[to_match]:
                to_match = lsp[to_match - 1]
            if string[end] == pattern[to_match]:
                to_match += 1
            if to_match == M:
                return end - M + 1
        
        return -1

    def get_lsp(pattern):
        N = len(pattern)
        lsp = [0] * N
        to_match = 0

        for end in range(1, N):
            while to_match > 0 and pattern[to_match] != pattern[end]:
                to_match = lsp[to_match - 1]
            if pattern[to_match] == pattern[end]:
                to_match += 1
            lsp[end] = to_match
        
        return lsp

    return matcher(elems, looking_for)

File: dsa/cpp/kmp.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // O(n + m)
    int kmp(const string& elems, const string& lookingFor) {
        return matcher(elems, lookingFor);
    }

private:
    vector<int> get_lsp(const string& pattern) {
        int N = (int)pattern.size();
        vector<int> lsp(N, 0);
        int mtc = 0;

        for (int i = 1; i < N; i++) {
            while (mtc > 0 && pattern[mtc] != pattern[i]) {
                mtc = lsp[mtc-1];
            }
            if (pattern[mtc] == pattern[i]) {
                mtc++;
            }
            lsp[i] = mtc;
        }

        return lsp;
    }

    int matcher(const string& s, const string& pattern) {
        int N = (int)s.size();
        int M = (int)pattern.size();
        vector<int> lsp = get_lsp(pattern);
        int mtc = 0;

        for (int i = 0; i < N; i++) {
            while (mtc > 0 && s[i] != pattern[mtc]) {
                mtc = lsp[mtc-1];
            }
            if (s[i] == pattern[mtc]) {
                mtc++;
            }
            if (mtc == M) {
                return i-M+1;
            }
        }

        return -1;
    }
};